### PR TITLE
Move class_typet::baset to struct_typet

### PR DESCRIPTION
**Pull request by tautschnig**
Structs can have base classes/structs. This enables wide use of the bases() API.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by kroening** [comment]
Does it make sense to have a middle layer -- struct_with_base_typet or so?
Most users of stuct_typet do not need protection or inheritance.

**Comment by tautschnig** [comment]
> Does it make sense to have a middle layer -- struct_with_base_typet or so?

Maybe - what is the cost of having this API in there, and thus the added value of `struct_with_base_typet`?

**Comment by smowton** [comment]
On the other hand does C++ really distinguish a class from a struct at all? I appreciate there are two keywords, but isn't a struct just a class with default public visibility, and the real distinction is between plain-old-data (C-ish structs) any anything else?

**Comment by tautschnig** [comment]
> On the other hand does C++ really distinguish a class from a struct at all?

I'd say the only difference is, as you said, the default access. Which has been clarified in #3002 (cf07c6c33).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: c6a55e73d92080f54810bbfbe73b76840b5dbda8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85748394

File: src/cpp/cpp_typecheck_bases.cpp [hunk 1]
```diff
@@ -141,19 +141,18 @@ void cpp_typecheckt::add_base_components(
     vbases.insert(from_name);
 
   // look at the the parents of the base type
-  forall_irep(it, from.find(ID_bases).get_sub())
+  for(const auto &b : from.bases())
   {
-    irep_idt sub_access=it->get(ID_access);
+    irep_idt sub_access = b.get(ID_access);
```
Comment (peterschrammel) [inline src/cpp/cpp_typecheck_bases.cpp:146]: Isn't there a `get_access()`?

File: src/cpp/cpp_typecheck_constructor.cpp [hunk 2]
```diff
@@ -496,12 +489,11 @@ void cpp_typecheckt::check_member_initializers(
...
       bool ok=false;
-      forall_irep(parent_it, bases.get_sub())
+      for(const auto &b : bases)
       {
-        assert(parent_it->get(ID_type) == ID_symbol_type);
-
-        if(member_type.get(ID_identifier)
-          ==parent_it->find(ID_type).get(ID_identifier))
+        if(
+          member_type.get(ID_identifier) ==
```
Comment (peterschrammel) [inline src/cpp/cpp_typecheck_constructor.cpp]: `get_identifier()` if possible?

**Review by peterschrammel — APPROVED** [review]

File: src/cpp/cpp_typecheck_bases.cpp (hunk 1 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_bases.cpp:146]: Not for a base (class), only for components.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 1f5e03cab0d578dcb69443dff20f74cf59cfe528).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85823211

**Pull request merged**